UDP Client: Split large points

joelegasse opened the pull request:
The v2 UDP client will attempt to split points that exceed the configured payload size. It will only do this for points that have a timestamp specified.

This follows from several ancestors: #6913 #7288 

###### Required for all non-trivial PRs
- [x] Rebased/mergable
- [x] Tests pass
- [x] CHANGELOG.md updated

###### Required only if applicable
_You can erase any checkboxes below this note if they are not applicable to your Pull Request._
- [ ] [InfluxQL Spec](https://github.com/influxdata/influxdb/blob/master/influxql/README.md) updated
- [ ] Provide example syntax
- [ ] Update man page when modifying a command

phemmer commented on client/v2/udp.go:
```diff
@@ -0,0 +1,123 @@
+	Addr string
+
+	// PayloadSize is the maximum size of a UDP client message, optional
+	// Tune this based on your network. Defaults to UDPBufferSize.
```
What is `UDPBufferSize`?

joelegasse replied:
A typo that should be updated to `UDPPayloadSize` :-)

phemmer commented on client/v2/udp.go:
```diff
@@ -0,0 +1,123 @@
+}
+
+// Ping will check to see if the server is up with an optional timeout on waiting for leader.
+// Ping returns how long the request took, the version of the server it connected to, and an error if one occurred.
```
Shouldnt the function description reflect what the function actually does?

joelegasse replied:
Absolutely, I'll remove this comment since it is not exported to the docs, and update the comment on `Ping` in the `Client` interface

phemmer commented on client/v2/udp.go:
```diff
@@ -0,0 +1,123 @@
+		if _, err := uc.conn.Write(b.Bytes()); err != nil {
+			return err
+		}
+	}
```
If we just wrote the buffer, shouldn't there be a `b.Reset()` here?

joelegasse replied:
We could, but the buffer is scoped to the current function, and the next line is to return.

phemmer replied:
Doh. Missed that. For some reason I thought the buffer persisted. Carry on :-)

phemmer commented on client/v2/udp.go:
```diff
@@ -0,0 +1,123 @@
+		if len(point) > MaxUDPPayloadSize {
+			delayedError = ErrLargePoint
+			continue
+		}
```
The size of the UDP payload is going to shrink after `p.pt.Split`, shouldn't this check come after we split it?

joelegasse replied:
Good point... I'll rework the logic to account for a `Split`-able point that by itself exceeds the max UDP payload size.

joelegasse replied:
Although, a single point at roughly 64KB would be an impressive amount of information to be storing in a single measurement at a single instant in time, but that doesn't mean we should prevent it from being written.

phemmer replied:
Also, what's the benefit of even checking the size prior to sending? Why not just try to write it, and handle the "write too large" error.

joelegasse replied:
Good point, I guess there isn't one. I'll remove the check altogether.

joelegasse commented:
My latest commit adds some improvements to the `models` package to reduce string allocations when writing points.

For now, these are only taken advantage of in the UDP client, but could be used by the HTTP client as well.

I also removed the `bytes.Buffer` and replaced it with a plain old byte slice.

joelegasse commented:
@phemmer @vlasad @Maksadbek @sebito91

You've all been involved with this issue for a while now. I think this current change is good enough to be merged. I would really appreciate any testing you could do with your data that would stress the point splitting behavior while ensuring correctness. I've added some changes that should increase performance over previous iterations: mostly memory re-use and in-place scanning, rather than the decode-reencode process that was happening previously.

I still need to update the docs for the client to describe the splitting behavior, along with the reasoning behind the various decisions we've come to over the past 2.5 months :-)

sebito91 commented:
@joelegasse thank you sir, we'll test this out now and get back to you ASAP.

jwilder commented:
@sebito91 Did you get a chance to test this out?

sebito91 commented:
Apologies fr the delay, but this works well for us!
